reward/reward_distributor.go

```diff
@@ -38,113 +50,456 @@ type governanceHelper interface {
 }
 
+// TODO: this is for legacy, will be removed
+type RewardDistributor struct{}
+
 func NewRewardDistributor(gh governanceHelper) *RewardDistributor {
-	return &RewardDistributor{
-		rcc: newRewardConfigCache(gh),
-		gh:  gh,
+	return &RewardDistributor{}
+}
+
+// DistributeBlockReward distributes a given block's reward at the end of block processing
+func DistributeBlockReward(b BalanceAdder, rewards map[common.Address]*big.Int) {
+	for addr, amount := range rewards {
+		b.AddBalance(addr, amount)
 	}
 }
 
-// getTotalTxFee returns the total transaction gas fee of the block.
-func (rd *RewardDistributor) getTotalTxFee(header *types.Header, rewardConfig *rewardConfig) *big.Int {
-	totalGasUsed := big.NewInt(0).SetUint64(header.GasUsed)
-	if header.BaseFee != nil {
-		// magma hardfork
-		return totalGasUsed.Mul(totalGasUsed, header.BaseFee)
+func NewRewardConfig(header *types.Header, config *params.ChainConfig) (*rewardConfig, error) {
+	if config.Governance == nil {
+		return nil, errors.New("no config.Governance")
+	}
+	if config.Governance.Reward == nil {
+		return nil, errors.New("no config.Governance.Reward")
+	}
+
+	rules := config.Rules(header.Number)
+
+	cnRatio, kgfRatio, kirRatio, totalRatio, err := parseRewardRatio(config.Governance.Reward.Ratio)
+	if err != nil {
+		return nil, err
+	}
+
+	var cnProposerRatio, cnStakingRatio, cnTotalRatio int64
+	if rules.IsKore {
+		cnProposerRatio, cnStakingRatio, cnTotalRatio, err = parseRewardKip82Ratio(config.Governance.Reward.Kip82Ratio)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &rewardConfig{
+		// hardfork rules
+		rules: rules,
+
+		// values calculated from block header
+		totalFee: GetTotalTxFee(header, config),
+
+		// values from ChainConfig
+		mintingAmount: new(big.Int).Set(config.Governance.Reward.MintingAmount),
+		minimumStake:  new(big.Int).Set(config.Governance.Reward.MinimumStake),
+		deferredTxFee: config.Governance.Reward.DeferredTxFee,
+
+		// parsed ratio
+		cnRatio:    big.NewInt(cnRatio),
+		kgfRatio:   big.NewInt(kgfRatio),
+		kirRatio:   big.NewInt(kirRatio),
+		totalRatio: big.NewInt(totalRatio),
+
+		// parsed KIP82 ratio
+		cnProposerRatio: big.NewInt(cnProposerRatio),
+		cnStakingRatio:  big.NewInt(cnStakingRatio),
+		cnTotalRatio:    big.NewInt(cnTotalRatio),
+	}, nil
+}
+
+func GetTotalTxFee(header *types.Header, config *params.ChainConfig) *big.Int {
+	totalFee := new(big.Int).SetUint64(header.GasUsed)
+	if config.IsMagmaForkEnabled(header.Number) {
+		totalFee = totalFee.Mul(totalFee, header.BaseFee)
 	} else {
-		return totalGasUsed.Mul(totalGasUsed, rewardConfig.unitPrice)
+		totalFee = totalFee.Mul(totalFee, new(big.Int).SetUint64(config.UnitPrice))
 	}
+	return totalFee
 }
 
-func (rd *RewardDistributor) txFeeBurning(txFee *big.Int) *big.Int {
-	return txFee.Div(txFee, big.NewInt(2))
+// GetBlockReward returns the actual reward amounts paid in this block
+// Used in klay_getReward RPC API
```

Comment:
I am sorry that I couldn't find this RPC API. Could you point the source line?

Reply:
It's not in this PR, but in https://github.com/klaytn/klaytn/pull/1666!

Reply:
Oops. Thanks your kind. 
I am sorry for late PR.